File: 022__Generate_Parentheses.py
class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        def addOne(list):
            temp = []
            for lst in list:
                for i in range(len(lst)):
                    cur = lst[:i] + "(" + lst[i:]
                    for j in range(i + 1, len(cur)):
                        current = cur[:j] + ")" + cur[j:]
                        if current not in temp:
                            temp.append(current)
            return temp
        result = ["()"]
        for i in range(n - 1):
            result = addOne(result)
        return result


# Backtracking

class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        ans = []
        def backtrack(S = '', left = 0, right = 0):
            if len(S) == 2 * N:
                ans.append(S)
                return
            if left < N:
                backtrack(S+'(', left+1, right)
            if right < left:
                backtrack(S+')', left, right+1)

        backtrack()
        return ans

File: Java/022__Generate_Parentheses.java
class Solution {
    List<String> result;
    public List<String> generateParenthesis(int n) {
        result = new ArrayList<>();
        helper("", n, 0, 0);
        return result;
    }

    private void helper(String cur, int n, int left, int right) {
        if (left > n || left < right) return;
        if (left == n && right == n) result.add(cur);
        helper(cur + "(", n, left + 1, right);
        helper(cur + ")", n, left, right + 1);
    }
}
